namespace-lock.go

```diff
@@ -69,3 +69,60 @@ func newNSLock() *nameSpaceLock {
 		count:   0,
 	}
 }
+
+// nameSpaceLockMap - name space lock map, provides primitives to LockNS and UnlockNS.
+type nameSpaceLockMap struct {
+	lockMap      map[nameSpaceParam]*nameSpaceLock
+	lockMapMutex *sync.Mutex
+}
+
+// Global name space lock.
+var nsMap *nameSpaceLockMap
+
+// initNSLockMap - initialize name space lock map.
+func initNSLockMap() {
+	nsMap = &nameSpaceLockMap{
+		lockMap:      make(map[nameSpaceParam]*nameSpaceLock),
+		lockMapMutex: &sync.Mutex{},
+	}
+}
+
+// LockNS - locks the given resource, using a previously allocated
+// name space lock or initializing a new one.
+func LockNS(volume, path string, readLock bool) {
+	nsMap.lockMapMutex.Lock()
+	defer nsMap.lockMapMutex.Unlock()
+
+	param := nameSpaceParam{volume, path}
+	nsLock, found := nsMap.lockMap[param]
+	if !found {
+		nsLock = newNSLock()
+	}
+
+	if readLock {
+		nsLock.RLock()
+	} else {
+		nsLock.Lock()
+	}
+
+	nsMap.lockMap[param] = nsLock
+}
+
+// UnlockNS - unlocks any previously acquired read or write locks.
+func UnlockNS(volume, path string, readLock bool) {
```

Comment:
readLock should not be an argument. Introduce RLock, RUnlock, Lock and Unlock methods..